notebook/notebookapp.py

```diff
@@ -1566,15 +1564,23 @@ def init_server_extensions(self):
         manager = ConfigManager(read_config_path=config_path)
         section = manager.get(self.config_file_name)
         extensions = section.get('NotebookApp', {}).get('nbserver_extensions', {})
+        
+        for modulename, enabled in sorted(extensions.items()):
+            if modulename not in self.nbserver_extensions:
```

Comment:
I think this inverts priority when config is defined at both conf.d and top-level. What we had before, was:

`if in nbserver_extensions and not in extensions, load it` (conf.d extensions has higher priority)

What we have here is:

`if in extensions and not in self.nbserver_extensions` (conf.d extensions has *lower* priority)

This resolves differently when a given config value is defined at both levels.

If you'd like to add a test that verifies this (would be super helpful!), try loading config where `jupyter_notebook_conifg.py` enables an extension and `jupyter_notebook_config.d/test.json` disables it. The result should be a disabled extension.

Reply:
when you say priority do you mean literal priority in terms of the Dict itself or in terms of overriding?

I had thought that user settings should override the defaults. 

If self.nbserver_extensions already has this modulename then we defer to the user's configuration loaded from the app. Otherwise we adhere to what we find in the `jupyter_notebook_config.d/` directory.

Reply:
> If you'd like to add a test that verifies this (would be super helpful!), try loading config where `jupyter_notebook_conifg.py` enables an extension and `jupyter_notebook_config.d/test.json` disables it.

That seems backwards to me, that means that as long as that library is installed I can never override its values.

Specifically in the case I'm imagining

`jupyter_notebook_conifg.py` disables an extension and `jupyter_notebook_config.d/test.json` enables the extension… If we always enable the extension there is no way for a user to override that library's default to enabling when it is installed.

Reply:
Where I think this gets murky is if a user's config ends up making their notebook servers not launchable by orchestrators like JupyterHub. As an operator I like to know that a deployment will reliably come up and if there are issues they get surfaced so that we can resolve them. In the current state of affairs like the inconsistency that's being addressed in this PR, the lack of config either doesn't surface (a jupyter server extension doesn't hard stop the server if improperly configured) or ends up buried away in the notebook server logs.

Reply:
> Where I think this gets murky is if a user's config ends up making their notebook servers not launchable by orchestrators like JupyterHub.

That feels like a different case in my mind than what is happening in the `jupyter_notebook_config.d` which is often not defined by deployers but by library maintainers as part of their data_files to their `setuptools.setup` call.

I would agree that there should be a case for some way to override user configuration available to deployers. But that shouldn’t be the same convention that libraries are using to ensure their extensions are loaded. 

Consider that if the person were to upgrade a library that autoenabled an extension by default (an extension that had been explicitly disabled by the deployed).  The user would then overwrite the configuration placed as part of deployment in the `jupyer_notebook_config.d`, and that would mean a user could inadvertently change deployment config without having any clue that it was happening.

Reply:
However I didn’t intend for this to be a change in behaviour so I’ll switch it back to the previous logic even if I disagree with that logic’s appropriateness to the purpose of this mechanism.

Reply:
Thanks for thinking through it. I think it's tricky to say whether conf.d vs top-level is 'user config' vs not. The priority was based on other conf.d-type systems (profile.d, etc.) where the top-level config is loaded first, followed by `.d/*`, thus giving .d higher priority. That said, in practice today, conf.d is typically written by installers, while config.{py|json} is written by user actions. It does indeed make sense for that to have highest priority in that case.

I'll go ahead and merge this one.